src/runtime/HalideRuntime.h

```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
 extern int halide_copy_to_device(void *user_context, struct halide_buffer_t *buf,
                                  const struct halide_device_interface_t *device_interface);
 
+/** Copy data from one buffer to another. The buffers may have
+ * different shapes and sizes, but the destination buffer's shape must
+ * be contained within the source buffer's shape. The source data is
+ * pulled from either device or host memory on the source, depending
```

Comment:
Another complication: if the call can deallocate a device handle, that's complicated as the handle could be a halide_device_malloc result, a combined device/host allocation, a wrapped_native, or a crop. This is already a bug in halide_copy_to_device when it transitions the buffer to a new device. I expect one can trigger this by passing an output buffer that is e.g. an OpenGL wrapped native, and having the output Func scheduled cuda....

Reply:
Hrm, everywhere we flip buffers from one device to another has the same problem - we just call device_free. Perhaps these should be errors instead?